File: src/tut01_nodes/solution/src/relay.cpp
// Copyright 2026 Jose Laruta

#include <algorithm>
#include <cctype>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

/// Re-publishes each `chatter` message upper-cased on `chatter_upper`.
class Relay : public rclcpp::Node
{
public:
  Relay()
  : Node("relay")
  {
    pub_ = create_publisher<std_msgs::msg::String>("chatter_upper", 10);
    sub_ = create_subscription<std_msgs::msg::String>(
      "chatter", 10, [this](const std_msgs::msg::String & msg) {on_message(msg);});
    RCLCPP_INFO(get_logger(), "relay started");
  }

private:
  /// Upper-case one message and publish it.
  void on_message(const std_msgs::msg::String & msg)
  {
    auto out = std_msgs::msg::String();
    out.data = msg.data;
    std::transform(
      out.data.begin(), out.data.end(), out.data.begin(),
      [](unsigned char c) {return std::toupper(c);});
    pub_->publish(out);
    RCLCPP_INFO(get_logger(), "relayed \"%s\"", out.data.c_str());
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Relay>());
  rclcpp::shutdown();
  return 0;
}

File: src/tut01_nodes/solution/tut01_nodes_solution/relay.py
# Copyright 2026 Jose Laruta
"""Exercise 1: re-publish `chatter` upper-cased on `chatter_upper`."""

import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from std_msgs.msg import String


class Relay(Node):
    """One node that is both a subscriber and a publisher."""

    def __init__(self) -> None:
        super().__init__('relay')
        self._pub = self.create_publisher(String, 'chatter_upper', 10)
        self._sub = self.create_subscription(String, 'chatter', self._on_message, 10)
        self.get_logger().info('relay started')

    def _on_message(self, msg: String) -> None:
        """Upper-case one message and publish it."""
        out = String()
        out.data = msg.data.upper()
        self._pub.publish(out)
        self.get_logger().info(f'relayed "{out.data}"')


def main(args: list[str] | None = None) -> None:
    """Entry point: init, spin, and shut down cleanly."""
    rclpy.init(args=args)
    node = Relay()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()
